src/foam/nanos/boot/NSpecFactory.java

```diff
@@ -91,15 +91,26 @@ void buildService(X x) {
     }
   }
 
-  public Object create(X x) {
-    synchronized (this) {
-      if ( ns_ == null ||
-           ns_ instanceof ProxyDAO && ((ProxyDAO) ns_).getDelegate() == null ) {
-        buildService(x);
+  private  ThreadLocal<Object> threadLocalNS_ = new ThreadLocal<Object>() {
+    @Override
+    protected Object initialValue() {
+      synchronized ( this ) {
+        if ( ns_ == null ) {
+          buildService(x_);
+        }
       }
+      return ns_;
     }
-    if ( ns_ instanceof XFactory ) return ((XFactory) ns_).create(x);
 
+    @Override
+    public Object get() {
+      return super.get();
+    }
+  };
+
+  public Object create(X x) {
+    threadLocalNS_.get();
```

Comment:
Contrary to earlier comments I made
```suggestion
    Object ns = threadLocalNS_.get();
```